### PR TITLE
Support prefix in @ConfigRoot

### LINKED ISSUE #9791: Add "prefix" to @ConfigRoot
**Description**
We are using Quarkus at a large scale to develop many microservices. In this setup we also have quite a few homegrown libraries and tools, which we now look to integrate into the Quarkus development as Quarkus extensions. For this purpose we want to use `@ConfigRoot` et al. to define the configuration properties and also generate corresponding documentation, as is being done for Quarkus itself. Since most of the extensions we are developing are not applicable outside our corporation, these won't be opensourced under the Quarkus project and so it doesn't really make any sense for the configuration properties to have the `quarkus` prefix.

I would therefor like to propose that the `@ConfigRoot` annotation type is extended with a `prefix` attribute (possibly defaulting to `"quarkus"`, so that we can specify our own prefix.

**Implementation ideas**
As described, I think a new `prefix` attribute on `@ConfigRoot` would make sense.

### PR DESCRIPTION
**Pull request by radcortez**
- Fixes https://github.com/quarkusio/quarkus/issues/9791

### CODE REVIEW

**Review by machi1990 — CHANGES_REQUESTED** [review]
Can we also do the necessary changes in the configuration doc generator? https://github.com/quarkusio/quarkus/blob/1eb14e58b0ffa84d56ab899978e43b976e8796ee/core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java#L79

**Comment by radcortez** [comment]
> Can we also do the necessary changes in the configuration doc generator?
> 
> https://github.com/quarkusio/quarkus/blob/1eb14e58b0ffa84d56ab899978e43b976e8796ee/core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java#L79

Sure.

**Comment by gsmet** [comment]
@radcortez what's the status of this? You're saying more tests are needed in the description. Did you add them, should we wait for them?

File: core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java
Comment (machi1990) [inline core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java:65]: What's the reasoning behind this change?

File: core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java
Comment (machi1990) [inline core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java:75]: Can be made a constant.

**Comment by radcortez** [comment]
> @radcortez what's the status of this? You're saying more tests are needed in the description. Did you add them, should we wait for them?

I want to add some more. It seems that we generally agree with the implementation, so let me complete with more tests. Thanks!

File: core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java
Comment (radcortez) [reply core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java:65]: Ops. Shouldn't be there. Was using these to test the generation with the test extension.

**Comment by radcortez** [comment]
Should be good to go if CI passed and we get an approval review :)

**Review by machi1990 — CHANGES_REQUESTED** [review]
Added some feedback on the doc generator.

**Comment by gsmet** [comment]
@machi1990 could you have another look?

**Review by machi1990 — APPROVED** [review]
LGTM, thanks  @radcortez

**Comment by knutwannheden** [comment]
I had completely missed this PR. Very cool!

A few notes / questions:

1. Did you consider `namespace` as an alternative to `prefix` (as discussed in the issue)?
2. Shouldn't also `ConfigDescriptionBuildStep#processConfig()` be adapted? See https://github.com/radcortez/quarkus/blob/c3218265101bd3d10d041c6eb692c8093f5f87cd/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigDescriptionBuildStep.java#L73
3. `ConfigGenerationBuildStep#checkForBuildTimeConfigChange()` appears to ignore the `RootDefinition`'s prefix: https://github.com/radcortez/quarkus/blob/configroot-prefix/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java#L115
4. Not sure if anything needs to be done about `ConfigUtils.BuildTimeSysPropConfigSource#getProperties()`: https://github.com/radcortez/quarkus/blob/configroot-prefix/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java#L242

I haven't tested the code yet, but I could look into it, if you think it would be helpful.

**Comment by radcortez** [comment]
Thanks for reviewing!

> I had completely missed this PR. Very cool!
> 
> A few notes / questions:
> 
> 1. Did you consider `namespace` as an alternative to `prefix` (as discussed in the issue)?

I think we should keep `prefix`. It was already used in `@ConfigProperties` and also in `@ConfigMapping`. Don't want to deviate too much. At some point I would also like to replace the backing implementation by the one used by `@ConfigMapping`. Right now, we have to maintain 3 different config flavours :(

> 2. Shouldn't also `ConfigDescriptionBuildStep#processConfig()` be adapted? See https://github.com/radcortez/quarkus/blob/c3218265101bd3d10d041c6eb692c8093f5f87cd/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigDescriptionBuildStep.java#L73

I'll have a look.

> 3. `ConfigGenerationBuildStep#checkForBuildTimeConfigChange()` appears to ignore the `RootDefinition`'s prefix: https://github.com/radcortez/quarkus/blob/configroot-prefix/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java#L115

I'll have a look.

> 4. Not sure if anything needs to be done about `ConfigUtils.BuildTimeSysPropConfigSource#getProperties()`: 
https://github.com/radcortez/quarkus/blob/configroot-prefix/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java#L242
> 

Maybe. I need to think about this.

> I haven't tested the code yet, but I could look into it, if you think it would be helpful.

Please, do! Thanks!

**Comment by knutwannheden** [comment]
I finally got around to test this PR with our own extensions. So far it all looks good. It would be very cool to get this integrated!

**Comment by knutwannheden** [comment]
Is there anything I can help with to make progress on this PR?

**Comment by radcortez** [comment]
> Is there anything I can help with to make progress on this PR?

I think this is going only after the Quarkus 2.0.0.Final release. We don't want to introduce major changes between the CR releases and the Final.

**Comment by knutwannheden** [comment]
Yes, that makes sense. I thought that 2.0 had already been branched. But it might of course make it more difficult to cherry-pick other changes if this gets integrated now.

Anyway, I will be happy to help with testing.

**Comment by radcortez** [comment]
Thanks. Right now, I just need to rebase due to a small conflict, but I don't expect any issues. The CI is passing, (the failing test is just a hiccup). Lets see how it behaves with the rebase.

**Comment by gsmet** [comment]
This needs a rebase and the failures in `native-config-profile` are suspicious.

Please ping me on Zulip once this is in better shape and it can be considered for merging.

**Comment by radcortez** [comment]
Sure.

**Comment by gastaldi** [comment]
What's the status of this PR? Is it ready to be merged?

**Review by gastaldi — APPROVED** [review]
I rebased this PR to make sure it is safe to be merged. It looks good to me BTW

**Comment by gastaldi** [comment]
I'll leave the decision to merge it to @gsmet and @radcortez 😉

**Comment by knutwannheden** [comment]
I think it would be great if this could make it into the 2.3 release. I can also offer to do some more manual testing if that is of any help.

**Comment by gastaldi** [comment]
Rebased again, let's merge this once CI is happy

**Comment by gastaldi** [comment]
And now this PR has some conflicts and requires rebasing. Can you have a look when possible @radcortez ? 

Thanks!

**Comment by radcortez** [comment]
> And now this PR has some conflicts and requires rebasing. Can you have a look when possible @radcortez ?
> 
> Thanks!

Done!

**Comment by gastaldi** [comment]
Failure is unrelated

**Pull request merged**
